nav2_pure_pursuit_controller/src/pure_pursuit_controller.cpp

```diff
@@ -123,34 +159,60 @@ geometry_msgs::msg::TwistStamped PurePursuitController::computeVelocityCommands(
 
 void PurePursuitController::setPlan(const nav_msgs::msg::Path & path)
 {
-  // Transform global path into the robot's frame
-  global_plan_ = transformGlobalPlan(path);
+  global_pub_->publish(path);
+  global_plan_ = path;
 }
 
-nav_msgs::msg::Path PurePursuitController::transformGlobalPlan(const nav_msgs::msg::Path & path)
+nav_msgs::msg::Path
+PurePursuitController::transformGlobalPlan(
+  const geometry_msgs::msg::PoseStamped & pose)
 {
   // Original mplementation taken fron nav2_dwb_controller
 
-  if (path.poses.empty()) {
+  if (global_plan_.poses.empty()) {
     throw nav2_core::PlannerException("Received plan with zero length");
   }
 
-  // Compute distance threshold of points on the plan that are outside the local costmap
+  // let's get the pose of the robot in the frame of the plan
+  geometry_msgs::msg::PoseStamped robot_pose;
+  if (!transformPose(
+      tf_, global_plan_.header.frame_id, pose,
+      robot_pose, transform_tolerance_))
+  {
+    throw nav2_core::PlannerException("Unable to transform robot pose into global plan's frame");
+  }
+
+  // we'll discard points on the plan that are outside the local costmap
   nav2_costmap_2d::Costmap2D * costmap = costmap_ros_->getCostmap();
-  double transform_dist_threshold =
-    std::max(costmap->getSizeInCellsX(), costmap->getSizeInCellsY()) *
+  double dist_threshold = std::max(costmap->getSizeInCellsX(), costmap->getSizeInCellsY()) *
     costmap->getResolution() / 2.0;
 
+  // First find the closest pose on the path to the robot
+  auto transformation_begin =
+    min_by(
+    global_plan_.poses.begin(), global_plan_.poses.end(),
+    [&robot_pose](const geometry_msgs::msg::PoseStamped & ps) {
+      return euclidean_distance(robot_pose, ps);
+    });
+
+
+  // Find the first pose in the end of the plan that's further than dist_threshold from the robot
```

Comment:
I'm not super clear what this line is supposed to mean. I can't decipher what its aim is.

I think at this point, you have the point in the path closest to the robot, lets call it at index `i`. The simplest thing to me would be to then work backwards from `i` and find the point `i - N` which is behind the robot's current pose on the path by more than 1 costmap cell (so `N` could be `0` or `1`) and then delete all points on the path before `i - N`. 

I see elements of that here but the comments don't make clear to me

Reply:
It looks like you're iterating from `i` to `end` (forward) and I don't know what you want to achieve with that. The over-use of lambdas here makes this very difficult to read.

Reply:
I think this probably also works, but I think the comments need to be updated so that its more immediately clear what direction you're taking to do this. I think there are multiple approaches so letting a reader know which you're using is good. I would think like find the earliest point in the path to keep, and then delete all before it. But I think instead you think like finding all the valid points first and then inverting the removal.

Reply:
It seems more straight forward to me to just look for `i - N`, remove all items before then, and then return the transformed path of `i - end`.

Reply:
These comments are adapted directly from [dwb_local_planner.cpp](https://github.com/ros-planning/navigation2/blob/2d31a8d3346d3a9ba6ab99a94d208bdf5879b413/nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp#L502). Once we figure out what the comments should say, I'll send a PR to fix them in DWB too.

That comment is indeed wrongly formulated. I've changed it to ` From the closest point, look for the first point that's further then dist_threshold from the robot. These points are definitly outside of the costmap so we won't transform them.` (see also the diff). See also the figure below. I've also reordered the code and changed several other comments.

`transformation_end` is probably an optimization that is needed for dwb. Points outside of the costmap will not be transformed, because DWB could never assign a cost to them.

For the pure pursuit controller this optimization is not needed. We could just transform from closest point to end. Should I change the code to that?

As for the style, I've used lambdas because dwb uses them. I thought that would result in the least review resistance.

![nav2 transform explained](https://user-images.githubusercontent.com/91969/99878030-b13c2e00-2c02-11eb-9117-f0c9c0e626ca.png)

Reply:
I'm sure DWB has its issues but nothing we can do about that right at this moment. It looks like my interpretation of that code was correct - I don't love it, but I think its fine for now. Not the optimization I would have made, but I think its OK. This will cause issues though if there are users that don't do replanning, but so will DWB so we'll wait until someone says its a problem to solve. The comment helps make it abundantly clear.

Reply:
Please also update the image to this new one - its a helpful detail.

Reply:
I've updated the image